### ISSUE TITLE
NullReferenceException

### ISSUE THREAD

**Issue by AdamCaviness** [richardschneider/net-mdns#79]
First off, thank you for all the work you put into this project; it is fantastic. 

I hook TaskScheduler.UnobservedTaskException and I occasionally get null ref exceptions as seen in the following log message. I'm running v2.0.1 (Makaretu.Dns.Multicast package v0.25.0).

```
   -------------Exception--------------
   Ex Type:      System.AggregateException
   Message:      A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread.
   Source:       
   Method:       
   Data:         
   Stack trace:  
                 ---------More Detail----------
                 AggregateException Inner Ex Count:               1
                 AggregateException Base Ex Type:                 System.NullReferenceException
                 AggregateException Base Ex Method:               Byte[] EndReceive(System.IAsyncResult, System.Net.IPEndPoint ByRef)
                 AggregateException Base Ex Message:              Object reference not set to an instance of an object.
                 AggregateException Base Ex Stack trace:          at System.Net.Sockets.UdpClient.EndReceive(IAsyncResult asyncResult, IPEndPoint& remoteEP)
                     at System.Net.Sockets.UdpClient.<ReceiveAsync>b__64_1(IAsyncResult ar)
                     at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar,
                     Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)
                     --- End of stack trace from previous location where
                     exception was thrown ---
                     at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
                     at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
                     at Makaretu.Dns.MulticastClient.<>c__DisplayClass13_0.<<Listen>b__0>d.MoveNext()
                 ---------Flattened Inner Exceptions 1/1---------
                 InnerException:                 
                     Object reference not set to an instance of an object.

      -------------Inner Exception 1--------------
      Inner Ex Type:     System.NullReferenceException
      Inner Message:     Object reference not set to an instance of an object.
      Inner Source:      System
      Inner Method:      Byte[] EndReceive(System.IAsyncResult, System.Net.IPEndPoint ByRef)
      Inner Data:        System.Collections.ListDictionaryInternal
      Inner Stack trace: at System.Net.Sockets.UdpClient.EndReceive(IAsyncResult asyncResult, IPEndPoint& remoteEP)
                         at System.Net.Sockets.UdpClient.<ReceiveAsync>b__64_1(IAsyncResult ar)
                         at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar,
                         Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)
                         --- End of stack trace from previous location where
                         exception was thrown ---
                         at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
                         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
                         at Makaretu.Dns.MulticastClient.<>c__DisplayClass13_0.<<Listen>b__0>d.MoveNext()
```

**Comment by richardschneider** [comment]
Thanks for the detailed bug report.

I suspect this is a race condition between Listen and Dispose in MulticastClient and is most likely caused by a  Network Interface change.

Does the app still reply to MDNS messages after you detect this error? I think it should.

**Comment by AdamCaviness** [comment]
I am only using it for service discovery only at the moment so I didn’t test after these exceptions (there were several of the same kind over a 15 minute span) and I shutdown my service and restarted it. To be clear, I’ve only hit this on our Web Api service, I have not seen it on a client yet. Our app isn’t released so I hit this on my dev machine which has a hyper-v switch that all my traffic goes through (100% out of the box hyper-v configuration). I also have Hamachi software vpn on it and occasionally connect via it. I do not believe Hamachi had been connected during that service lifetime but I clearly recall quitting the Hamachi app during the lifetime which may have quickly executed some unknown interface changes.

**Comment by richardschneider** [comment]
The `UdpClient` does not allow cancellation of `ReceiveAsync`, see this [issue](https://github.com/dotnet/corefx/issues/9848).  So when a NIC changes, the library must dispose the `UdpClient`.

I'll change the receive code to also catch the `NullReferenceException`.

**Comment by richardschneider** [comment]
Fixed in release v0.25.1

**Comment by Luai-m** [comment]
I occasionally get this unhandled exception when I have attached the  TaskScheduler.UnobservedTaskException:

 

```2019-10-16 16:17:02.429 +03:00 (14732:0002) [ERR] AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread.

   SocketException: A request to send or receive data was disallowed because the socket had already been shut down in that direction with a previous shutdown call
   at System.Net.Sockets.Socket.DoBeginReceiveFrom(Byte[], Int32, Int32, System.Net.Sockets.SocketFlags, System.Net.EndPoint, System.Net.SocketAddress, System.Net.Sockets.OverlappedAsyncResult);System.dll;100671645;212
   at System.Net.Sockets.Socket.BeginReceiveFrom(Byte[], Int32, Int32, System.Net.Sockets.SocketFlags, System.Net.EndPoint ByRef, System.AsyncCallback, System.Object);System.dll;100671644;270
   at System.Net.Sockets.UdpClient.BeginReceive(System.AsyncCallback, System.Object);System.dll;100671973;42
   at System.Threading.Tasks.TaskFactory`1.FromAsyncImpl(System.Func`3[System.AsyncCallback,System.Object,System.IAsyncResult], System.Func`2[System.IAsyncResult,TResult], System.Action`1[System.IAsyncResult], System.Object, System.Threading.Tasks.TaskCreationOptions);mscorlib.dll;100679367;304
   at System.Net.Sockets.UdpClient.ReceiveAsync();System.dll;100671984;17
   at Makaretu.Dns.MulticastClient+<>c__DisplayClass13_0+<<Listen>b__0>d.MoveNext();Makaretu.Dns.Multicast.dll;100663321;23```

@richardschneider does #81 also fix this. If not maybe the receive code should catch all exceptions?
